src/analysis/asymptotics/superpolynomial_decay.lean

```diff
@@ -13,296 +13,288 @@ import data.polynomial.eval
-  (hα : tendsto (λ x, ∥algebra_map α 𝕜 x∥) at_top at_top) :
-  superpolynomial_decay f ↔
-    ∀ (c : ℤ), is_o f (λ x, (algebra_map α 𝕜 x) ^ c) at_top :=
-begin
-  refine ⟨λ h c, _, λ h c, (h c).is_O⟩,
-  have hα' : ∀ᶠ (x : α) in at_top, (algebra_map α 𝕜 x) ≠ 0,
-  from (eventually_ne_of_tendsto_norm_at_top hα 0).mono (λ x hx hx', absurd hx' hx),
-  have : is_o (λ x, 1 : α → 𝕜) (λ x, (algebra_map α 𝕜 x)) at_top,
-  { refine is_o_of_tendsto' (hα'.mono $ λ x hx hx', absurd hx' hx)
-      (tendsto_zero_iff_norm_tendsto_zero.2 _),
-    simp only [one_div, normed_field.norm_inv],
-    exact tendsto.comp tendsto_inv_at_top_zero hα },
-  have := this.mul_is_O (h $ c - 1),
-  simp only [one_mul] at this,
-  refine this.trans_is_O (is_O.of_bound 1 (hα'.mono (λ x hx, le_of_eq _))),
-  rw [zpow_sub_one₀ hx, mul_comm, mul_assoc, inv_mul_cancel hx, one_mul, mul_one]
-end
+lemma superpolynomial_decay.congr' (hf : superpolynomial_decay l k f)
+  (hfg : f =ᶠ[l] g) : superpolynomial_decay l k g :=
+λ z, (hf z).congr' (eventually_eq.mul (eventually_eq.refl l _) hfg)
+
+lemma superpolynomial_decay.congr (hf : superpolynomial_decay l k f)
+  (hfg : ∀ x, f x = g x) : superpolynomial_decay l k g :=
+λ z, (hf z).congr (λ x, congr_arg (λ a, k x ^ z * a) $ hfg x)
+
+@[simp]
+lemma superpolynomial_decay_zero (l : filter α) (k : α → β) :
+  superpolynomial_decay l k 0 :=
+λ z, by simpa only [pi.zero_apply, mul_zero] using tendsto_const_nhds
+
+lemma superpolynomial_decay.add [has_continuous_add β] (hf : superpolynomial_decay l k f)
+  (hg : superpolynomial_decay l k g) : superpolynomial_decay l k (f + g) :=
+λ z, by simpa only [mul_add, add_zero, pi.add_apply] using (hf z).add (hg z)
 
-theorem superpolynomial_decay_iff_norm_tendsto_zero (f : α → 𝕜)
-  (hα : tendsto (λ x, ∥algebra_map α 𝕜 x∥) at_top at_top) :
-  superpolynomial_decay f ↔
-    ∀ (c : ℤ), tendsto (λ x, ∥f x * (algebra_map α 𝕜 x) ^ c∥) at_top (𝓝 0) :=
+lemma superpolynomial_decay.mul [has_continuous_mul β] (hf : superpolynomial_decay l k f)
+  (hg : superpolynomial_decay l k g) : superpolynomial_decay l k (f * g) :=
+λ z, by simpa only [mul_assoc, one_mul, mul_zero, pow_zero] using (hf z).mul (hg 0)
+
+lemma superpolynomial_decay.mul_const [has_continuous_mul β] (hf : superpolynomial_decay l k f)
+  (c : β) : superpolynomial_decay l k (λ n, f n * c) :=
+λ z, by simpa only [←mul_assoc, zero_mul] using tendsto.mul_const c (hf z)
+
+lemma superpolynomial_decay.const_mul [has_continuous_mul β] (hf : superpolynomial_decay l k f)
+  (c : β) : superpolynomial_decay l k (λ n, c * f n) :=
+(hf.mul_const c).congr (λ _, mul_comm _ _)
+
+lemma superpolynomial_decay.parameter_mul (hf : superpolynomial_decay l k f) :
+  superpolynomial_decay l k (k * f) :=
+λ z, tendsto_nhds.2 (λ s hs hs0, l.sets_of_superset ((tendsto_nhds.1 (hf $ z + 1)) s hs hs0)
+  (λ x hx, by simpa only [set.mem_preimage, pi.mul_apply, ← mul_assoc, ← pow_succ'] using hx))
+
+lemma superpolynomial_decay.mul_parameter (hf : superpolynomial_decay l k f) :
```

Comment:
What do you think of renaming `parameter` to `param` in these names? A bit shorter, just as readable.

Reply:
I think that makes a lot of sense, I've changed all these names